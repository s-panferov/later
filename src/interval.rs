use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;

#[derive(Debug, Clone)]
pub struct Interval {
	pub from: DateTime<Utc>,
	pub to: Option<DateTime<Utc>>,
}

impl Interval {
	pub fn from_now() -> Self {
		Interval {
			from: Utc::now(),
			to: None,
		}
	}

	pub fn from(from: DateTime<Utc>) -> Self {
		Interval { from, to: None }
	}
}

/// Something that resolves to the range of time
pub trait AsIntervals: Debug {
	/// Get the estimate of the interval
	fn duration_hint(&self) -> Duration;

	/// Get an interator to resolve intervals
	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>>;
}
