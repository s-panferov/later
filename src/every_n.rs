use chrono::Duration;

use crate::interval::{AsIntervals, Interval};

#[derive(Debug)]
pub struct EveryN(Duration, i32);

impl EveryN {
	pub fn new(d: Duration, n: i32) -> EveryN {
		EveryN(d, n)
	}
}

impl AsIntervals for EveryN {
	fn duration(&self) -> &Duration {
		&self.0
	}

	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>> {
		Box::new(EveryNIterator {
			duration: self.0.clone(),
			interval,
			n: self.1,
		})
	}
}

struct EveryNIterator {
	duration: Duration,
	interval: Interval,
	n: i32,
}

impl Iterator for EveryNIterator {
	type Item = Interval;
	fn next(&mut self) -> Option<Interval> {
		let from = self.interval.from.clone();
		let to = self.interval.from + self.duration;

		if self.interval.to.is_some()
			&& to >= *self.interval.to.as_ref().unwrap()
		{
			return None;
		}

		self.interval.from = to + self.duration * (self.n - 1);
		Some(Interval { from, to: Some(to) })
	}
}
