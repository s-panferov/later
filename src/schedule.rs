use crate::interval::{AsIntervals, Interval};

pub enum SchedulePart {
	Every(Box<dyn AsIntervals>),
}

pub struct Schedule {
	parts: Vec<SchedulePart>,
}

impl Schedule {
	pub fn from_parts(parts: Vec<SchedulePart>) -> Self {
		Schedule { parts }
	}

	/// Get an interator to resolve intervals
	pub fn iter_within(&self, interval: Interval) -> ScheduleIterator {
		ScheduleIterator::new(self, interval)
	}
}

pub struct ScheduleIterator<'a> {
	interval: Interval,
	schedule: &'a Schedule,
	every: Vec<&'a Box<dyn AsIntervals>>,
	initialized: bool,
	state: Vec<Box<dyn Iterator<Item = Interval>>>,
}

impl<'a> ScheduleIterator<'a> {
	fn new(schedule: &'a Schedule, interval: Interval) -> Self {
		let mut every = vec![];
		for p in &schedule.parts {
			match p {
				SchedulePart::Every(ref e) => every.push(e),
			}
		}

		every.sort_by(|a, b| b.duration().cmp(&a.duration()));

		ScheduleIterator {
			schedule,
			every,
			interval,
			initialized: false,
			state: Vec::new(),
		}
	}
}

impl<'a> ScheduleIterator<'a> {
	fn init(&mut self, mut i: usize, interval: Interval) -> Option<Interval> {
		let mut int = interval.clone();
		let len = self.every.len();
		while i < len {
			let mut iter = self.every.get(i).unwrap().iter_within(int.clone());
			let next = iter.next();
			match next {
				Some(next) => {
					int = next;
					self.state.push(iter);
				}
				None => return None,
			}
			i += 1;
		}

		Some(int)
	}
}

impl<'a> Iterator for ScheduleIterator<'a> {
	type Item = Interval;

	fn next(&mut self) -> Option<Interval> {
		if !self.initialized {
			let len = self.every.len();
			if len == 0 {
				return None;
			}

			let int = self.init(0, self.interval.clone());
			match int {
				Some(int) => {
					self.initialized = true;
					return Some(int);
				}
				None => return None,
			}
		}

		let curr_iter = self.state.last_mut().unwrap();
		let next = curr_iter.next();

		match next {
			Some(next) => return Some(next),
			None => loop {
				self.state.pop();
				let len = self.state.len();
				if len == 0 {
					return None;
				}
				let curr_iter = self.state.last_mut().unwrap();
				let next = curr_iter.next();
				match next {
					Some(next) => return self.init(len, next),
					None => continue,
				}
			},
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::every_n::EveryN;
	use chrono::Duration;

	use insta::assert_debug_snapshot_matches;

	#[test]
	fn every() {
		let schedule = Schedule::from_parts(vec![
			SchedulePart::Every(Box::new(Duration::minutes(10))),
			SchedulePart::Every(Box::new(EveryN::new(Duration::hours(1), 2))),
		]);

		let what: Vec<Interval> = schedule
			.iter_within(Interval::from(
				"2019-01-01T00:00:00Z".parse().unwrap(),
			))
			.take(1000)
			.collect();

		assert_debug_snapshot_matches!("every", what);
	}
}
