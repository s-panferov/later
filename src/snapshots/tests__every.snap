---
created: "2019-08-31T23:26:51.551581Z"
creator: insta@0.10.1
source: src/schedule.rs
expression: what
---
[
    Interval {
        from: 2019-01-01T00:00:00Z,
        to: Some(
            2019-01-01T00:10:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T00:10:00Z,
        to: Some(
            2019-01-01T00:20:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T00:20:00Z,
        to: Some(
            2019-01-01T00:30:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T00:30:00Z,
        to: Some(
            2019-01-01T00:40:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T00:40:00Z,
        to: Some(
            2019-01-01T00:50:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T00:50:00Z,
        to: Some(
            2019-01-01T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T02:00:00Z,
        to: Some(
            2019-01-01T02:10:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T02:10:00Z,
        to: Some(
            2019-01-01T02:20:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T02:20:00Z,
        to: Some(
            2019-01-01T02:30:00Z,
        ),
    },
    Interval {
        from: 2019-01-01T02:30:00Z,
        to: Some(
            2019-01-01T02:40:00Z,
        ),
    },
]
